File: combinatorics/resource/TMD/graph_tree/Code/Problem05/exercise1.6.py
from typing import List, Tuple, Iterator

# A type alias for a matching. It's a list of tuples, where each
# tuple is a matched pair (x_vertex, y_vertex).
Matching = List[Tuple[int, int]]

def enumerate_perfect_matchings(p: int, q: int) -> Iterator[Matching]:
    """
    Exercise 1.6: Enumerates all perfect matchings in a complete bipartite graph K_p,q.

    A perfect matching pairs every vertex from the smaller partition with a unique
    vertex from the larger one. For a true "perfect matching" covering ALL vertices
    of the graph, we must have p == q. This function handles the general definition
    and will only yield results if p == q.

    This implementation uses a recursive backtracking algorithm to generate all
    permutations of pairings.

    :param p: The size of the first partition (X).
    :param q: The size of the second partition (Y).
    :return: A generator that yields each perfect matching as a list of tuples.
    """
    # A perfect matching that covers all vertices can only exist if p == q.
    if p != q or p <= 0:
        return
        yield # An empty generator

    # Vertices in partition Y that are still available to be matched.
    available_y = list(range(p))
    
    def find_matchings_recursive(current_matching: Matching, remaining_y: List[int]):
        """
        A recursive helper function to find matchings via backtracking.

        :param current_matching: The list of pairs matched so far.
        :param remaining_y: The list of vertices in partition Y still available.
        """
        # Base case: if no vertices are left in Y, we have found a complete matching.
        if not remaining_y:
            yield current_matching
            return

        # Determine the next vertex from partition X to match.
        # Its index is simply the number of pairs we've already matched.
        x_vertex_idx = len(current_matching)

        # Try matching x_vertex_idx with each available vertex in Y.
        for i in range(len(remaining_y)):
            y_vertex = remaining_y[i]
            
            # Create a new list of remaining Y vertices for the recursive call.
            next_remaining_y = remaining_y[:i] + remaining_y[i+1:]
            
            # Recurse: add the new pair and pass the smaller list of available vertices.
            yield from find_matchings_recursive(
                current_matching + [(x_vertex_idx, y_vertex)],
                next_remaining_y
            )

    # Start the recursion with an empty matching and all Y vertices available.
    yield from find_matchings_recursive([], available_y)

def print_matchings(matchings_generator: Iterator[Matching], p: int, q: int):
    """Helper function to print the list of matchings."""
    matchings = list(matchings_generator)
    print(f"--- Found {len(matchings)} perfect matchings for K_{p},{q} ---")
    if not matchings and p != q:
        print("  (As expected, no perfect matching exists since p != q)")
    elif not matchings and p == 0:
        print("  (An empty matching for an empty graph)")
    else:
        for i, matching in enumerate(matchings, 1):
            edges_str = ", ".join([f"(x{x}, y{y})" for x, y in matching])
            print(f"  Matching {i}: {{ {edges_str} }}")
    print("------------------------------------------")

# =================================================================================
# MAIN EXECUTION BLOCK - DEMONSTRATION FOR EXERCISE 1.6
# =================================================================================
if __name__ == "__main__":
    print("=" * 60)
    print(" Perfect Matching Enumeration Demo (Exercise 1.6)    ")
    print("=" * 60)

    # --- Case 1: p = q, perfect matchings exist ---
    p1, q1 = 3, 3
    # The number of matchings should be 3! = 6
    matchings1 = enumerate_perfect_matchings(p1, q1)
    print_matchings(matchings1, p1, q1)

    # --- Case 2: p != q, no perfect matchings ---
    p2, q2 = 3, 2
    matchings2 = enumerate_perfect_matchings(p2, q2)
    print_matchings(matchings2, p2, q2)
    
    # --- Case 3: p = q = 4 ---
    p3, q3 = 4, 4
    # The number of matchings should be 4! = 24
    matchings3 = enumerate_perfect_matchings(p3, q3)
    print_matchings(matchings3, p3, q3)

    print("=" * 60)

File: combinatorics/resource/TMD/graph_tree/Code/Problem05/exercise1.6.cpp
#include <iostream>
#include <vector>
#include <numeric>     // For std::iota
#include <algorithm>   // For std::next_permutation
#include <string>

// A type alias for a matching, which is a vector of pairs.
// Each pair represents a matched edge, e.g., {(x_vertex, y_vertex), ...}.
using Matching = std::vector<std::pair<int, int>>;

/**
 * @brief Exercise 1.6: Enumerates all perfect matchings in a complete bipartite graph K_p,p.
 *
 * A perfect matching in K_p,p pairs every vertex from one partition (say, X of size p)
 * with a unique vertex from the other partition (Y of size p). The number of such
 * matchings is p!, which corresponds to the number of permutations of the vertices in Y.
 *
 * This function generates all these permutations to list every possible perfect matching.
 * It returns an empty list if p != q, as perfect matchings do not exist.
 *
 * @param p The size of the first partition.
 * @param q The size of the second partition.
 * @return A vector containing all possible perfect matchings.
 */
std::vector<Matching> enumerate_perfect_matchings_cpp(int p, int q) {
    // A perfect matching can only exist if the partitions have the same size.
    if (p != q || p <= 0) {
        return {};
    }
    
    std::vector<Matching> all_matchings;
    
    // Create a vector representing the vertices in partition Y: {0, 1, ..., p-1}.
    std::vector<int> partition_Y(p);
    std::iota(partition_Y.begin(), partition_Y.end(), 0);

    // Use std::next_permutation to generate all possible pairings for partition Y.
    // The vertices in partition X are implicitly {0, 1, ..., p-1}.
    do {
        Matching current_matching;
        // For each permutation of Y, create the corresponding matching.
        for (int i = 0; i < p; ++i) {
            // Match vertex i from partition X with vertex partition_Y[i] from partition Y.
            current_matching.push_back({i, partition_Y[i]});
        }
        all_matchings.push_back(current_matching);
    } while (std::next_permutation(partition_Y.begin(), partition_Y.end()));

    return all_matchings;
}

/**
 * @brief Helper function to print the list of matchings.
 */
void print_matchings(const std::vector<Matching>& all_matchings, int p, int q) {
    std::cout << "--- Found " << all_matchings.size() << " perfect matchings for K_" 
              << p << "," << q << " ---" << std::endl;
    if (all_matchings.empty() && p != q) {
        std::cout << "  (As expected, no perfect matching exists since p != q)" << std::endl;
    } else {
        int count = 1;
        for (const auto& matching : all_matchings) {
            std::cout << "  Matching " << count++ << ": { ";
            for (const auto& edge : matching) {
                std::cout << "(x" << edge.first << ", y" << edge.second << ") ";
            }
            std::cout << "}" << std::endl;
        }
    }
    std::cout << "------------------------------------------" << std::endl;
}

// =================================================================================
// MAIN FUNCTION - DEMONSTRATION FOR EXERCISE 1.6
// =================================================================================
int main() {
    std::cout << "=====================================================" << std::endl;
    std::cout << " Perfect Matching Enumeration Demo (Exercise 1.6)    " << std::endl;
    std::cout << "=====================================================" << std::endl;

    // --- Case 1: p = q, perfect matchings exist ---
    int p1 = 3, q1 = 3;
    std::vector<Matching> matchings1 = enumerate_perfect_matchings_cpp(p1, q1);
    print_matchings(matchings1, p1, q1);

    // --- Case 2: p != q, no perfect matchings ---
    int p2 = 3, q2 = 2;
    std::vector<Matching> matchings2 = enumerate_perfect_matchings_cpp(p2, q2);
    print_matchings(matchings2, p2, q2);
    
    // --- Case 3: p = q = 0, edge case ---
    int p3 = 0, q3 = 0;
    std::vector<Matching> matchings3 = enumerate_perfect_matchings_cpp(p3, q3);
    print_matchings(matchings3, p3, q3);

    std::cout << "=====================================================" << std::endl;

    return 0;
}